jax/lax/lax_control_flow.py

```diff
@@ -264,29 +264,29 @@ def while_loop(cond_fun, body_fun, init_val):
       # transformation on it), so we fall back to the primitive version.
       pass
 
-  # The body input and output avals must match exactly. However, we want to account for
-  # the case when init contains weakly-typed values (e.g. Python scalars), with avals that
-  # may not match the output despite being compatible by virtue of their weak type.
-  # To do this, we compute the jaxpr in two passes: first with the raw inputs, and if
-  # necessary, a second time with modified init values.
-  for pass_ in ["first", "second"]:
+  def _create_jaxpr(init_val):
     init_vals, in_tree = tree_flatten((init_val,))
     init_avals = tuple(_map(_abstractify, init_vals))
     cond_jaxpr, cond_consts, cond_tree = _initial_style_jaxpr(cond_fun, in_tree, init_avals)
     body_jaxpr, body_consts, body_tree = _initial_style_jaxpr(body_fun, in_tree, init_avals)
     if not treedef_is_leaf(cond_tree) or len(cond_jaxpr.out_avals) != 1:
-      msg = "cond_fun must return a boolean scalar, but got pytree {} in {} pass."
-      raise TypeError(msg.format(cond_tree, pass_))
+      msg = "cond_fun must return a boolean scalar, but got pytree {}."
+      raise TypeError(msg.format(cond_tree))
     if cond_jaxpr.out_avals[0].strip_weak_type() != ShapedArray((), np.bool_):
-      msg = "cond_fun must return a boolean scalar, but got output type(s) {} in {} pass."
-      raise TypeError(msg.format(cond_jaxpr.out_avals, pass_))
+      msg = "cond_fun must return a boolean scalar, but got output type(s) {}."
+      raise TypeError(msg.format(cond_jaxpr.out_avals))
+    return init_vals, in_tree, init_avals, cond_jaxpr, cond_consts, body_jaxpr, body_consts, body_tree
 
-    if pass_ != "first":
-      break
-    new_init_vals, changed = _promote_weak_typed_inputs(init_vals, init_avals, body_jaxpr.out_avals)
-    if not changed:
-      break
-    init_val, = tree_unflatten(in_tree, new_init_vals)
+  # The body input and output avals must match exactly. However, we want to account for
+  # the case when init contains weakly-typed values (e.g. Python scalars), with avals that
+  # may not match the output despite being compatible by virtue of their weak type.
+  # To do this, we compute the jaxpr in two passes: first with the raw inputs, and if
+  # necessary, a second time with modified init values.
+  init_vals, in_tree, init_avals, cond_jaxpr, cond_consts, body_jaxpr, body_consts, body_tree = _create_jaxpr(init_val)
```

Comment:
To save a little horizontal space we could re-arrange the returned tuple and write this as
```python
init_vals, init_avals, body_jaxpr, *rest = _create_jaxpr(init_val)
new_init_vals, changed = _promote_weak_typed_inputs(init_vals, init_avals, body_jaxpr.out_avals)
if changed:
  new_init_val, = tree_unflatten(in_tree, new_init_vals)
  init_vals, init_avals, body_jaxpr, *rest = _create_jaxpr(new_init_val)
in_tree, cond_jaxpr, cond_consts, body_consts, body_tree = rest
```

We could also consider recomputing stuff if it's too annoying to plumb out.

Anyway we always end up revising code over time, and this code'll be no different, so I suggest just finishing off a version you're happy with and moving on :)

Reply:
I like that idea. Done.